eval/bundled/epm.elv.go

```diff
@@ -1,111 +1,279 @@
 package bundled
 
-const epmElv = `
+const epmElv = `# Verbosity configuration
+debug-mode = $false
+
+# Configuration for common domains
+-default-domain-config = [
+  &"github.com"= [
+    &method= git
+    &protocol= https
+    &levels= 2
+  ]
+  &"bitbucket.org"= [
+    &method= git
+    &protocol= https
+    &levels= 2
+  ]
+  &"gitlab.com"= [
+    &method= git
+    &protocol= https
+    &levels= 2
+  ]
+]
+
+# Internal configuration
 -data-dir = ~/.elvish
--installed = $-data-dir/epm-installed
+-lib-dir = $-data-dir/lib
 
-fn -info [text]{
-    print (edit:styled '=> ' green)
+# Runtime state - records get copied from -default-domain-config
+# or read from epm-domain.cfg files as needed
+-domain-config = [&]
+
+# General utility functions
+
+fn -debug [text]{
+  if $debug-mode {
+    print (edit:styled '=> ' blue)
     echo $text
+  }
+}
+
+fn -info [text]{
+  print (edit:styled '=> ' green)
+  echo $text
+}
+
+fn -warn [text]{
+  print (edit:styled '=> ' yellow)
+  echo $text
 }
 
 fn -error [text]{
-    print (edit:styled '=> ' red)
-    echo $text
+  print (edit:styled '=> ' red)
+  echo $text
 }
 
-fn add-installed [@pkgs]{
-    if (eq $pkgs []) {
-        -error 'Must specify at least one package.'
-        return
-    }
-    for pkg $pkgs {
-        echo $pkg >> $-installed
-    }
+fn dest [pkg]{
+  put $-lib-dir/$pkg
 }
 
-fn -get-url [pkg]{
-    put https://$pkg
+fn is-installed [pkg]{
+  put ?(test -e (dest $pkg))
 }
 
-fn -install-one [pkg]{
-    dest = $-data-dir/lib/$pkg
-    if ?(test -e $dest) {
-        -error 'Package '$pkg' already exists locally.'
-        return
-    }
-    -info 'Installing '$pkg
-    mkdir -p $dest
-    git clone (-get-url $pkg) $dest
-    add-installed $pkg
+fn -package-domain [pkg]{
+  splits / $pkg | take 1
 }
 
-fn install [@pkgs]{
-    if (eq $pkgs []) {
-        -error 'Must specify at least one package.'
-        return
+fn -package-without-domain [pkg]{
+  joins / [(splits / $pkg | drop 1)]
```

Comment:
This can be: `splits / $pkg | drop 1 | joins /`

You can also avoid unnecessary splitting beyond the first / by

`splits &max=2 $pkg | drop 1 | joins ''`

You still need the last `joins ''` to correctly handle package names that do not contain any slashes (although I wonder whether that case is handled correctly everywhere else.)

Reply:
Thanks for the tip about `splits | ... | joins`!

The package-without-slashes case is handled correctly - I have locally an `rsync` package with a single level, package names like `dev/epm`, and they work fine. The minimum is `domain/package`.

Reply:
I was thinking of root-level packages, like just `zzamboni.org`. It's a question whether such packages should be supported at all. If we don't, we should check that package names contain slashes.

If we know that the package name will always contain at least one slash, the trailing `| join ''` will not be necessary.